include/nbl/asset/utils/CVertexHashGrid.h

```diff
@@ -0,0 +1,212 @@
+	{
+		if (vertexCount < (0x1ull << 16ull))
+			return core::LSBSorter<KeyAccessor::key_bit_count, uint16_t>();
+		if (vertexCount < (0x1ull << 32ull))
+			return core::LSBSorter<KeyAccessor::key_bit_count, uint32_t>();
+		return core::LSBSorter<KeyAccessor::key_bit_count, size_t>();
+	}
+
+	collection_t m_vertices;
+	const uint32_t m_hashTableMaxSize;
+	const float m_cellSize;
+
+	uint32_t hash(const VertexData& vertex) const
+	{
+		const hlsl::float32_t3 position = floor(vertex.getPosition() / m_cellSize);
+
+		return	((static_cast<uint32_t>(position.x) * primeNumber1) ^
+			(static_cast<uint32_t>(position.y) * primeNumber2) ^
+			(static_cast<uint32_t>(position.z) * primeNumber3))& (m_hashTableMaxSize - 1);
+	}
+
+	uint32_t hash(const hlsl::uint32_t3& position) const
+	{
+		return	((position.x * primeNumber1) ^
+			(position.y * primeNumber2) ^
+			(position.z * primeNumber3))& (m_hashTableMaxSize - 1);
+	}
+
+	uint8_t getNeighboringCellHashes(uint32_t* outNeighbors, const VertexData& vertex) const
+	{
+		hlsl::float32_t3 cellfloatcoord = floor(vertex.getPosition() / m_cellSize - hlsl::float32_t3(0.5));
+		hlsl::uint32_t3 baseCoord = hlsl::uint32_t3(static_cast<uint32_t>(cellfloatcoord.x), static_cast<uint32_t>(cellfloatcoord.y), static_cast<uint32_t>(cellfloatcoord.z));
+
+		uint8_t neighborCount = 0;
+
+		outNeighbors[neighborCount] = hash(baseCoord);
+		neighborCount++;
+
+		auto addUniqueNeighbor = [&neighborCount, outNeighbors](uint32_t hashval)
+		{
+			if (std::find(outNeighbors, outNeighbors + neighborCount, hashval) == outNeighbors + neighborCount)
+			{
+				outNeighbors[neighborCount] = hashval;
+				neighborCount++;
+			}
+		};
+
+		addUniqueNeighbor(hash(baseCoord + hlsl::uint32_t3(0, 0, 1)));
+		addUniqueNeighbor(hash(baseCoord + hlsl::uint32_t3(0, 1, 0)));
+		addUniqueNeighbor(hash(baseCoord + hlsl::uint32_t3(1, 0, 0)));
+		addUniqueNeighbor(hash(baseCoord + hlsl::uint32_t3(1, 1, 0)));
+		addUniqueNeighbor(hash(baseCoord + hlsl::uint32_t3(1, 0, 1)));
+		addUniqueNeighbor(hash(baseCoord + hlsl::uint32_t3(0, 1, 1)));
+		addUniqueNeighbor(hash(baseCoord + hlsl::uint32_t3(1, 1, 1)));
+
+		return neighborCount;
+	}
+
+	BucketBounds getBucketBoundsByHash(uint32_t hash) const
+	{
+		if (hash == invalidHash)
+			return { m_vertices.end(), m_vertices.end() };
+
+		const auto skipListBound = std::visit([&](auto& sorter)
+		{
+			auto hashBound = sorter.getHashBound(hash);
+			return std::pair<collection_t::const_iterator, collection_t::const_iterator>(m_vertices.begin() + hashBound.first, m_vertices.begin() + hashBound.second);
+		}, m_sorter);
+
+		auto begin = std::lower_bound(
+			skipListBound.first, 
+			skipListBound.second, 
+			hash,
+			[](const VertexData& vertex, uint32_t hash)
+			{
+				return vertex.hash < hash;
+			});
+
+		auto end = std::upper_bound(
+			skipListBound.first, 
+			skipListBound.second, 
+			hash, 
+			[](uint32_t hash, const VertexData& vertex)
+			{
+				return hash < vertex.hash;
```

Comment:
nitpick, if you have a `setHash`, then also have a `getHash` on the vertex, opens up the door to choosing to recompute the hash every `getHash` and not storing it in the vertex